azure-auth-helper/src/main/java/com/microsoft/azure/auth/LocalAuthServer.java

```diff
@@ -0,0 +1,123 @@
+    private static String loginSuccessBodyTemplate;
+    private static String loginErrorBodyTemplate;
+
+    private HttpServer server;
+    private String code;
+    private String error;
+    private String errorDescription;
+    private boolean finish;
+
+
+    public LocalAuthServer() throws IOException {
+        initHtmlTemplate();
+        server = HttpServer.create(new InetSocketAddress(0), 0);
+        server.createContext("/", httpExchange -> {
+            try {
+                final Map<String, String> attributeMap = QueryStringUtil.queryToMap(httpExchange.getRequestURI().getQuery());
+                LocalAuthServer.this.code = attributeMap.get(CODE);
+                LocalAuthServer.this.error = attributeMap.get(ERROR);
+                LocalAuthServer.this.errorDescription = attributeMap.containsKey("error_description") ? attributeMap.get("error_description")
+                        : attributeMap.get("error_response");
+
+                final boolean isSuccess = StringUtils.isEmpty(error) && StringUtils.isNotEmpty(code);
+
+
+                final OutputStreamWriter osw = new OutputStreamWriter(httpExchange.getResponseBody());
+                if (isSuccess) {
+                    httpExchange.sendResponseHeaders(200, loginSuccessBodyTemplate.length());
+                    osw.write(loginSuccessBodyTemplate);
+                    osw.flush();
+                } else {
+                    final String response = String.format(loginErrorBodyTemplate, error, errorDescription);
+                    httpExchange.sendResponseHeaders(200, response.length());
+                    osw.write(response);
+                    osw.flush();
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
+            } finally {
+                server.stop(0);
+                finish = true;
+            }
+        });
+        server.setExecutor(null); // creates a default executor
+    }
+
+    public int getPort() {
+        return server.getAddress().getPort();
+    }
+
+    public String getUrl() {
+        return "http://localhost:" + getPort();
+    }
+
+    public void start() {
+        server.start();
+    }
+
+    public void stop() {
+        server.stop(0);
+    }
+
+    public String getResult() throws InterruptedException, AzureLoginFailureException {
+        while (!finish) {
+            Thread.sleep(50);
+        }
+        if (StringUtils.isEmpty(error) && StringUtils.isNotEmpty(code)) {
+            return code;
+        }
+        if (StringUtils.isNotEmpty(error)) {
+            if (StringUtils.isNotEmpty(errorDescription)) {
+                throw new AzureLoginFailureException(error + "\nReason: " + errorDescription);
+            }
+            throw new AzureLoginFailureException(error);
+        } else {
+            throw new AzureLoginFailureException("There is no error and no code.");
+        }
+    }
+
+    static void initHtmlTemplate() throws IOException {
```

Comment:
Shall we use a static block to do the init job?

Reply:
A lazy init will be better